### PR TITLE
Metatensor interface for vesin

### PR DESCRIPTION
**Pull request by ceriottm**
This should somehow reduce the boilerplate code when using vesin with metatensor.

### CODE REVIEW

File: python/vesin-torch/vesin/torch/metatensor/_neighbors.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    def compute(
+        self,
+        system: System,
+    ) -> TensorBlock:
+        """
+        Compute the neighbor list for the metatensor.System ``system`` and adds
+        to its neighbor lists
+
+        :param system: a :py:class:`System
+            <metatensor.torch.atomistic.System>`_ object containing the
```
Comment (Luthaf) [inline python/vesin-torch/vesin/torch/metatensor/_neighbors.py:33]: This link looks weird, did you checked that it resolves to the right thing? Or do we need to add intersphinx here?

File: python/vesin-torch/vesin/torch/metatensor/_neighbors.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            data about a structure.
+
+        :return: a :py:class:`metatensor.TensorBlock` containing
+            the neighbor list data. The data will also be added to
+            the set of neighbor lists associated with ``system``.
```
Comment (Luthaf) [inline python/vesin-torch/vesin/torch/metatensor/_neighbors.py:38]: I would either add the nl to the system or return it, but doing both feels confusing.

File: python/vesin-torch/vesin/torch/metatensor/_neighbors.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        else:
+            raise NotImplementedError("`vesin` currently does not support mixed pbcs.")
+        
+        # computes neighbor list
+        (
+            i,
+            j,
+            S,
+            D,
+        ) = self._nl._c.compute(
```
Comment (Luthaf) [inline python/vesin-torch/vesin/torch/metatensor/_neighbors.py:56]: why are you accessing `._c` here? This should work fine with the public API

File: python/vesin-torch/vesin/torch/metatensor/_neighbors.py (hunk 1 above)
Comment (ceriottm) [reply python/vesin-torch/vesin/torch/metatensor/_neighbors.py]: probably needs intersphinx

Comment (ceriottm) [reply python/vesin-torch/vesin/torch/metatensor/_neighbors.py]: ok I think I fixed it

File: python/vesin-torch/vesin/torch/metatensor/_neighbors.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                ],
+                values=torch.hstack([i.reshape(-1, 1), j.reshape(-1, 1), S]),
+            ),
+            components=[Labels.range("xyz", 3)],
+            properties=Labels(names=["distance"], values=torch.tensor([[0]])),
+        )
+
+        # store into the System and return
+        if system.positions.requires_grad or system.cell.requires_grad:
+            register_autograd_neighbors(system, nl_block)
```
Comment (Luthaf) [inline python/vesin-torch/vesin/torch/metatensor/_neighbors.py:135]: I don't think we need this one, since this repo contains the same hack already and uses it by default for the `vesin.torch` calculator. Worth adding a small gradcheck test though. Maybe just trying to compute the gradient of NL w.r.t. positions, which will fail if we are missing a part of the graph?

File: python/vesin-torch/vesin/torch/metatensor/_neighbors.py (hunk 2 above)
Comment (Luthaf) [reply python/vesin-torch/vesin/torch/metatensor/_neighbors.py:89]: Different people should be using this code vs calling `get_neighbor_list`, so I would rather keep these separate.

**Review by Luthaf — COMMENTED** [review]
I'm thinking a bit more about the API and use case here. Is the idea to use this when implementing a custom engine (which could be just the tests of a specific model), or something else? 

In general, since a model can have multiple NL, maybe we should offer to compute all of them? So keeping the current code to do `NeighborListOptions` => `TensorBlock` for a single `System`, and adding a separate function that would look like

```py
def setup_neighbors(model: MetatensorAtomisticModel, systems: List[System])
```

That would (a) get the NL options from the model, (b) compute all NL for all suystems and (c) add the NL to the systems in-place.

**Comment by ceriottm** [comment]
> I'm thinking a bit more about the API and use case here. Is the idea to use this when implementing a custom engine (which could be just the tests of a specific model), or something else?
> 
> In general, since a model can have multiple NL, maybe we should offer to compute all of them? So keeping the current code to do `NeighborListOptions` => `TensorBlock` for a single `System`, and adding a separate function that would look like
> 
> ```python
> def setup_neighbors(model: MetatensorAtomisticModel, systems: List[System])
> ```
> 
> That would (a) get the NL options from the model, (b) compute all NL for all suystems and (c) add the NL to the systems in-place.

The idea is to avoid all the boilerplate code of calling vesin, then packing the result in a tensorblock, then add it to the system. I would say that higher functionalities like the one you suggest here should be taken care of by the model engine, so maybe implemented MTA-side? at the same time it does need to know about the specific NL implementation so maybe you're right we can also have this in here.

File: python/vesin-torch/vesin/torch/metatensor/_neighbors.py (hunk 4 above)
Comment (ceriottm) [reply python/vesin-torch/vesin/torch/metatensor/_neighbors.py]: fixed and added test

**Comment by ceriottm** [comment]
> > I would say that higher functionalities like the one you suggest here should be taken care of by the model engine, so maybe implemented MTA-side?
> 
> So metatensor-atomistic does not provide NL, this is the responsibility of the engine. If the point here is to make it easier to write new engines (I count writing tests for metatensor models as a small, custom engine), then I think it make sense to provide this kind of functionality in vesin.
> 
> Overall, I'm not very clear on what's your use case here, apart from better integration between the different pieces of the ecosystem. Do you want to use this in torch-pme tests, elsewhere, ... ?

Can be used for tests, for quick toy models and for prototyping. Basically every time one would use the ase neighbor list, we can use this without needing to wrap the system into an ase atoms or whatever.
